pin.go

```diff
@@ -0,0 +1,168 @@
+package httpapi
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/interface-go-ipfs-core"
+	caopts "github.com/ipfs/interface-go-ipfs-core/options"
+	"github.com/pkg/errors"
+)
+
+type PinAPI HttpApi
+
+type pinRefKeyObject struct {
+	Type string
+}
+
+type pinRefKeyList struct {
+	Keys map[string]pinRefKeyObject
+}
+
+type pin struct {
+	path iface.ResolvedPath
+	typ  string
+}
+
+func (p *pin) Path() iface.ResolvedPath {
+	return p.path
+}
+
+func (p *pin) Type() string {
+	return p.typ
+}
+
+func (api *PinAPI) Add(ctx context.Context, p iface.Path, opts ...caopts.PinAddOption) error {
+	options, err := caopts.PinAddOptions(opts...)
+	if err != nil {
+		return err
+	}
+
+	return api.core().request("pin/add", p.String()).
+		Option("recursive", options.Recursive).Exec(ctx, nil)
+}
+
+func (api *PinAPI) Ls(ctx context.Context, opts ...caopts.PinLsOption) ([]iface.Pin, error) {
+	options, err := caopts.PinLsOptions(opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	var out pinRefKeyList
+	err = api.core().request("pin/ls").
+		Option("type", options.Type).Exec(ctx, &out)
+	if err != nil {
+		return nil, err
+	}
+
+	pins := make([]iface.Pin, 0, len(out.Keys))
+	for hash, p := range out.Keys {
+		c, err := cid.Parse(hash)
+		if err != nil {
+			return nil, err
+		}
+		pins = append(pins, &pin{typ: p.Type, path: iface.IpldPath(c)})
+	}
+
+	return pins, nil
+}
+
+func (api *PinAPI) Rm(ctx context.Context, p iface.Path, opts ...caopts.PinRmOption) error {
+	options, err := caopts.PinRmOptions(opts...)
+	if err != nil {
+		return err
+	}
+
+	return api.core().request("pin/rm", p.String()).
+		Option("recursive", options.Recursive).
+		Exec(ctx, nil)
+}
+
+func (api *PinAPI) Update(ctx context.Context, from iface.Path, to iface.Path, opts ...caopts.PinUpdateOption) error {
+	options, err := caopts.PinUpdateOptions(opts...)
+	if err != nil {
+		return err
+	}
+
+	return api.core().request("pin/update").
+		Option("unpin", options.Unpin).Exec(ctx, nil)
+}
+
+type pinVerifyRes struct {
```

Comment:
I'd consider using "temporary" structs instead of reusing the same type but it doesn't really matter much either way. I'm just not a fan of dead fields.

We can also just punt on this.